README.md
# ts-form-validation

[![Greenkeeper badge](https://badges.greenkeeper.io/mindhivefi/ts-form-validation.svg)](https://greenkeeper.io/)

This is a simple library to do form validation using Typescript's magnificent strong typing features. Framework focuses in giving an uniform architecture for validating a form fields and a form as a whole with a proper state validation. To do actual field type validations, it is recommended to use this library besides some other library such as [validator -library](https://www.npmjs.com/package/validator).

The framework is been kept in simple, so that it should be easy to integrate into a different frameworks. The framework is been tested in production use with React and React Native projects.

## Installation

All you need to do is to:

```bash
yarn add ts-form-validation
```

or

```bash
npm install ts-form-validation
```

## Migrating from 1.x version

In 2.0 field validator's validate method have a more inituitive notation:

```typescript
    value =>
        // return true if field is valid or if not return an error message object:
        validator.isLength(value, { min: 5, max: 30 }) || {
          type: MessageType.ERROR,
          message: 'Display name must be between 5 to 30 characters in length.',
        },
```

The old notation which was returning false when there was no need for a message, continues to work but
it is encouraged to update the form validation to a new notation.

## Using validator

The whole validation is based on a type interface defined by you, which will contain all the fields that are used in your form. This might be easier to explain with an example. Lets build a registration form:

### 1. Define the type for the form

Define an interface which will contain all fields that are used in your form:

```typescript
interface RegisterForm {
  displayName: string;
  email: string;
  photoURL?: string;
  password1: string;
  password2: string;
}
```

### 2. Define rules for your fields

You define all the form login on rules object. Rules object is defined based on your interface created before:

```typescript
const rules: FormValidationRules<RegisterForm> {
  fields: {
    displayName: {
      required: true,
      trim: true,
      validate: (value: string) =>
        validator.isLength(value, { min: 5, max: 30 }) || {
          type: MessageType.ERROR,
          message: 'Display name must be between 5 to 30 characters in length.',
        },
    },
    email: {
      required: true,
      trim: true,
      // We are using validator -library here to do the actual email -check
      validate: (value: string) =>
        validator.isEmail(value) || {
          type: MessageType.ERROR,
          message: 'Please give a valid email address',
        },
    },
    password1: {
      required: true,
      validate: (value: string) => validator.isLength(value, {min: 7, max: 20}) || {
          type: MessageType.ERROR,
          message: 'Password must be at least 7-20 characters in length',
        },
    },
    password2: {
      required: true,
      validate: (value: any) => validator.isLength(value, {min: 7, max: 20}) || {
          type: MessageType.ERROR,
          message: 'Password must be at least 7-20 characters in length',
        },
    },
  }
};
```

Now we have the basic checking rules for form fields. All fields except the `photoURL` -field are required. Notice that `validator.isEmail` and `validator.isLength` -functions are imported from validator -library, which is specialized to check correctnes of a single values. You can use validator library, any other checker or do your own checkers with `ts-form-validation`. This framework only helps you with the structure and strong typing.

### 3. Validation rule to check interdependant rules on form

On registration form, it is important that the user will write the same password twice. This requires the code to match values between to different fields. This can be done with `validateForm` -event handler. So we will add a handler for that to rules object:

```typescript
const rules: FormValidationRules<RegisterForm> {
  validateForm: form => {
    const messages = {};

    // Do not make checking if user has not yet filled both fields
    if (form.filled.password1 && form.filled.password2 && form.values.password1 !== form.values.password2) {
      return {
        ...form,
        messages,
        formMessage: {
          type: MessageType.ERROR,
          message: 'Password do not match',
        },
      };
    }
    return { ...form, messages };
  },
  // Field definitions continue here
  fields: {
    displayName: {
      required: true,
      trim: true,
    },
    ...
  }
};
```

The `validateForm` -handler will be called after all field validators have been triggered. The method must return `false`, when form validation will not find anything to notice. When there are some messages that we must inform to user, we can set a message for specific field on messages -object with the same key that the field has. In this example we will set an error message for both password -fields. There is also `formMessage` -field, where you can locale a general message that is targeted for the whole form.

The message interface supports three kind of messages:

```typescript
export interface ValidationMessage {
  /**
   * Type of the error message.
   */
  type: MessageType;

  /**
   * Message code if any available
   */
  code?: FormValidationMessageCode;
  /**
   * The actual message
   */
  message: string;
}
```

Currently supported message types are INFO, WARNING and ERROR. It up to you how you use them in the user interface.

### 4. Adding preprocessing for form fields

It is possible to do basic trimming for input or add your own pre processor events for fields. A common case that is needed for form input is to trim white spaces away from start and the end of the text. This can be done simply adding a `trim: true`-field for a field that must be trimmed. We already used this feature in the example above like this:

```typescript
  email: {
    required: true,
    trim: true,
    validate: (value: string) =>
        validator.isEmail(value) || {
          type: MessageType.ERROR,
          message: 'Please give a valid email address',
        },
```

To add a custom preprocessor, just add event `preprocess`-event handler to field like this:

```typescript
  currency: {
    required: true,
    trim: true,
    preprocess: (value: string) => value.toUppercase(),
  },
```

You can use trim and custom event handler together. When you use trim also, the value that your custom event handler will receive will already be trimmed.

### 5. Filled - letting the user to finnish writing before validating

So that we will not make the user irritated, we must be able the define when a field is ready to be validated. For that reason, this framwwork has a `filled` -set on form definition. When you know that the user has given an input that is ready to be validated, you just set a form field's key to true in this set. After that the validation will be executed for that fields on each call. Normally this will be set to true, in form field's `onBlur`-event handled which will happend when user will leave the field.

## Example using validation with React and Material-ui

With react, its natural to make an own component for the form. Let's make a component to support the register form that we created above:

```typescript
interface State extends WithForm { // <- this will include form -property to state
  
}

export default class RegisterScreen extends React.Component<any, State> {
  public state: State = {
    // Initialize form
    form: initForm<RegisterForm>(
      {
        displayName: '',
        email: '',
        password1: '',
        password2: '',
        photoURL: '',
      },
      rules,
    ),
  };

  public render() {
    const { classes } = this.props;
    const { formMessage, isFormValid } = this.state.form;

    return (
      <Paper className={classes.container}>
        <Typography variant="h3" gutterBottom>
          Register form example
        </Typography>
        {formMessage && (
          <FormHelperText error={formMessage.type === MessageType.ERROR}>
            {formMessage.message}
          </FormHelperText>
        )}
        {this.renderField('displayName', 'Display name')}
        {this.renderField('email', 'Email')}
        {this.renderField('password1', 'Password', { type: 'password' })}
        {this.renderField('password2', 'Password again', { type: 'password' })}

        <Button
          color="primary"
          variant="contained"
          disabled={!isFormValid}
          onClick={() => alert('Form ready to go!')}
        >
          Register
        </Button>
      </Paper>
    );
  }

  /**
   * Render a single form input field with all spices
   */
  private renderField = (
    key: keyof RegisterForm,
    label: string,
    props?: Partial<TextFieldProps>,
  ) => {
    const { classes } = this.props;

    const {
      values,
      messages,
    } = this.state.form;
    return (
      <>
        <TextField
          {...props}
          id={key}
          label={label}
          className={classes.textField}
          value={values[key]}
          onBlur={this.handleBlur(key)}
          onChange={this.handleChange(key)}
          margin="normal"
          variant="outlined"
        />
        {message && (
          <FormHelperText error={message.type === MessageType.ERROR}>
            {message.message}
          </FormHelperText>
        )}
      </>
    );
  };

  private handleChange = (key: keyof RegisterForm) => (
    event: React.ChangeEvent<HTMLInputElement>,
  ) => {

    this.setState({
      form: validateForm(
        this.state.form, {
        // Disable preprocess while validating when writing
        usePreprocessor: false,
        // Update the field value
        setValues: {
          [key] event.target.value
        }
      }),
    });
  };

  private handleBlur = (key: keyof RegisterForm) => (
    event: React.ChangeEvent<HTMLInputElement>,
  ) => {
    // set field filled after blur, means that the field as been set once
    this.setState({
      form: validateForm(this.state.form, {
        setFilled: key, // mark the field be filled
      }),
    });
  };
}
```

## Api

### InputForm

InputForm is the object to wrap up all the form functionality.

| Property | Description |
|----------|-------------|
| values: &lt;T&gt;   | The current field values on the form presented as an object where each key is the name of field and value is the field value. |
| rules: FormValidationRules&lt;T&gt;   | Rules object containing the validation and constraints of the form |
| filled?: FilledFormFields&lt;T&gt;   | Marked for each field that have a value to show possible errors on field. Fields will be validated even when the field is not set but no error message will be used. In most of the cases you do not implicitly update filled values. This cab be let be done by `validateForm` -function. |

### FieldValidator

Field validator is an validator interface for a single fields validation.

| Property | Description |
|----------|-------------|
| required: boolean   | When set to true, the field value must not be undefined. If the field is not defined. The required text will shown. |
| requiredText: string \| MessageFunction | Own text to shown when required field have no value after validation. This requires that the filled of the field is also set to true. |
| trim?: boolean | Trims field's value from white spaces before and after the value. If this is set to true, this will be done before `preprosess` -method |
| preprocess? (value: T) => T | Do custom preprocessing for the field before the validation. If trim is true, the value given as argument will be trimmed on this phase. Just return the preprocessed value as return value and it will be used by validate -method and formValidation. |
| validate? (value: T) => T | Do validation for the field. When field is valid after validation the return value must be `true`. When the value is erronous, the result value must be `ValidationMessage` -object. |

### ValidationMessage

ValidationMessage describes an message to be shown to the user when there are some issues that the user should be aware of.

| Property | Description |
|----------|-------------|
| type: MessageType | Type of the validation message  | 
| |  - |
| message: string | Human readable message to desribe the user what is the issue with the field or the whole form. | 

### MessageType

ValidationMessage describes an message to be shown to the user when there are some issues that the user should be aware of.

| Key | Description |
|----------|-------------|
| ERROR | Field or form contains an error. If any validator, will return a message with this code, the form will be treated as invalid (isFormValid will be false). |
| WARNING | Field or form contains a state for what user should be warned. Form can still to be valid event if it contains warnings. |
| HINT | Field or form contains a state for what user should be hinted. Form can be treaded valid when containing this kind of messages. |
| VALIDATION_ERROR | This validation error is to help to find errors at development phase. Error message given, when validator callback fails with the exception. Exception error message will be included into message.  |

## Helpers

### initForm

With init form, you can initialize form with required fields, so that you can set up the minimun required:

```typescript
initForm<T>(
      {
        // initial values for fields
      },
      rules,
    ),

```

### WithForm

`WithForm` is an interface to make it easy to add form property to State interface

```typescript
interface State extends WithForm<FormType> {
}
```

Is same as:

```typescript
interface State {
  form: Form<FormType>;
}
```

### Custom messages for validation

Currently validator has only a single internal message that is given as result. This will hapen when you use ´required´ on
form validation rules. To customize this, you can add ´requiredText´-field into field validator:

```typescript
currency: {
    required: true,
    requiredText: 'You must give the currency code here',
    trim: true,
    preprocess: (value: string) => value.toUppercase(),
  },
```

If your system have multiple localizations, you might prefer to set a function into requiredText, which will give you a full control over message to be shown:

```typescript
currency: {
    required: true,
    requiredText: () => 'You must give the currency code here',
    trim: true,
    preprocess: (value: string) => value.toUppercase(),
  },
```

To set a new default message to be used in all fields that do not have explicit text override like done above, you can set default
message texts or message functions on rule -objects top level:

```typescript
const rules: FormValidationRules<RegisterForm> {
  ...
  defaultMessages: {
    requiredField: () => 'Hey, you missed this field!';
  };
}
```

### formHaveMessagesOfType

After validation, you can user `formHaveMessagesOfType`-function to check if the listed messages has some sort of messages, like:

```typescript
if formHaveMessagesOfType(form, MessageType.WARNING) {
  // do something handy
}
```

Please check [fully functional demo](https://github.com/mindhivefi/ts-form-validation-demo). for this example in its own repo.


CHANGELOG.md
## [2.1.1](https://github.com/mindhivefi/ts-form-validation/compare/v2.1.0...v2.1.1) (2021-05-09)


### Bug Fixes

* fix for a validateForm issue which did override  isFormValid state set by field validators ([2904ce8](https://github.com/mindhivefi/ts-form-validation/commit/2904ce81dc25614d2bbf9bba700bc7118d08c1eb))

# [2.1.0](https://github.com/mindhivefi/ts-form-validation/compare/v2.0.3...v2.1.0) (2021-04-05)


### Bug Fixes

* temproraly removed condition-circle out of use ([dee6586](https://github.com/mindhivefi/ts-form-validation/commit/dee6586e7a187e2c8b2f4a283923bccdc064e08a))
* version conflicts with the develop branch ([c92d922](https://github.com/mindhivefi/ts-form-validation/commit/c92d922153c23365c019e05f5046d9cfeddac5d7))


### Features

* streamlined api for form level validation ([fffc96a](https://github.com/mindhivefi/ts-form-validation/commit/fffc96a2017b58c98a8c0d1dadd1d634d54629c2))

## [2.0.3](https://github.com/mindhivefi/ts-form-validation/compare/v2.0.2...v2.0.3) (2019-07-21)


### Bug Fixes

* set up usePreprocessor as a default value for options ([36b95d9](https://github.com/mindhivefi/ts-form-validation/commit/36b95d9))

## [2.0.2](https://github.com/mindhivefi/ts-form-validation/compare/v2.0.1...v2.0.2) (2019-05-26)


### Bug Fixes

* Fixed interface use on test case + more test cases ([8138c87](https://github.com/mindhivefi/ts-form-validation/commit/8138c87))

## [2.0.1](https://github.com/mindhivefi/ts-form-validation/compare/v2.0.0...v2.0.1) (2019-05-23)


### Bug Fixes

* Changed validate and preprocess value argument to optional ([4c666e0](https://github.com/mindhivefi/ts-form-validation/commit/4c666e0))

# [2.0.0](https://github.com/mindhivefi/ts-form-validation/compare/v1.2.1...v2.0.0) (2019-05-22)


### Bug Fixes

* Fixed an issue with an example on documentation ([fba0aa5](https://github.com/mindhivefi/ts-form-validation/commit/fba0aa5))
* Fixed erronous generic param type on form interface ([51efc6f](https://github.com/mindhivefi/ts-form-validation/commit/51efc6f))


### Features

* Streamlined api for validation ([f0ec0a4](https://github.com/mindhivefi/ts-form-validation/commit/f0ec0a4))


### BREAKING CHANGES

* validate -functions notation changed. The old notation still works but will cause
warning on validate.

## [1.2.1](https://github.com/mindhivefi/ts-form-validation/compare/v1.2.0...v1.2.1) (2019-05-22)


### Bug Fixes

* Correction to WithForm documentation on readme ([126973a](https://github.com/mindhivefi/ts-form-validation/commit/126973a))

# [1.2.0](https://github.com/mindhivefi/ts-form-validation/compare/v1.1.6...v1.2.0) (2019-05-22)


### Features

* WithForm helper interface to streamline setting the form to state ([dc91361](https://github.com/mindhivefi/ts-form-validation/commit/dc91361))

## [1.1.6](https://github.com/mindhivefi/ts-form-validation/compare/v1.1.5...v1.1.6) (2019-03-26)


### Bug Fixes

* home-page-link-changed-to-wiki ([164c486](https://github.com/mindhivefi/ts-form-validation/commit/164c486))

## [1.1.5](https://github.com/mindhivefi/ts-form-validation/compare/v1.1.4...v1.1.5) (2019-03-15)


### Bug Fixes

* commonjs-module ([8e2a4bd](https://github.com/mindhivefi/ts-form-validation/commit/8e2a4bd))

## [1.1.4](https://github.com/mindhivefi/ts-form-validation/compare/v1.1.3...v1.1.4) (2019-03-14)


### Bug Fixes

* testing-lib ([3e46c31](https://github.com/mindhivefi/ts-form-validation/commit/3e46c31))

## [1.1.3](https://github.com/mindhivefi/ts-form-validation/compare/v1.1.2...v1.1.3) (2019-03-14)


### Bug Fixes

* optimized-package-publishing-size ([526dbe0](https://github.com/mindhivefi/ts-form-validation/commit/526dbe0))

## [1.1.2](https://github.com/mindhivefi/ts-form-validation/compare/v1.1.1...v1.1.2) (2019-03-14)


### Bug Fixes

* force-rebuild-on-deploy ([c145cb9](https://github.com/mindhivefi/ts-form-validation/commit/c145cb9))

## [1.1.1](https://github.com/mindhivefi/ts-form-validation/compare/v1.1.0...v1.1.1) (2019-03-13)


### Bug Fixes

* fix-gitignore-from-excluding-lib ([272bca8](https://github.com/mindhivefi/ts-form-validation/commit/272bca8))

# [1.1.0](https://github.com/mindhivefi/ts-form-validation/compare/v1.0.0...v1.1.0) (2019-03-13)


### Features

* customizable-required-field-messages ([6a30587](https://github.com/mindhivefi/ts-form-validation/commit/6a30587))

# 1.0.0 (2019-03-13)


### Bug Fixes

* semantice-release-fix ([8dfea44](https://github.com/mindhivefi/ts-form-validation/commit/8dfea44))


lib/index.d.ts
/**
 * # ts-form-validation
 *
 * A simple form validation library for typescript users.
 *
 * @author Ville Venäläinen / Mindhive Oy
 */
export declare const DEFAULT_ERRORMESSAGE_FIELD_IS_REQUIRED = "This field is required";
/**
 * Validator definitions for a single form field.
 */
export interface FieldValidator {
    /**
     * Is field required. Default is false.
     */
    required?: boolean;
    /**
     * Message text to be shown when field has no value but is required. If this field has no value,
     * the default value set on rules will be used or if that is not available the internal default
     * value will be shown.
     */
    requiredText?: string | MessageFunction;
    /**
     * Trim whitespaces from string's begining and end.
     */
    trim?: boolean;
    /**
     * Do custom preprocessing for the form value, before the validation.
     */
    preprocess?: (value: any) => any;
    /**
     * Validate Fields given value.
     *
     * @returns {false | ValidationMessage} Validate if the field should have some level of message. If no validation message is given, the method must return ´false´.
     */
    validate?: (value: any) => false | ValidationMessage;
}
/**
 * Initialize form with default set
 */
export declare const initForm: <T>(values: T, rules: FormValidationRules<T>) => Form<T>;
export interface Form<T> extends InputForm<T>, ValidatedForm<T> {
}
/**
 * Definition of the form to be validated
 */
export interface InputForm<T> {
    /**
     * Current form values
     *
     * @type {T}
     * @memberof Form
     */
    values: T;
    /**
     * Set of fields that user has already filled an input. If input is not given, the
     * field will not be validated.
     *
     * @type {FilledFormFields<T>}
     * @memberof Form
     */
    filled: FilledFormFields<T>;
    /**
     * Rules for the validation.
     *
     * @type {FormValidationRules<T>}
     * @memberof Form
     */
    rules: FormValidationRules<T>;
}
declare type BooleanMap<T> = {
    [P in keyof T]?: boolean;
};
declare type MapType<T, U> = {
    [P in keyof T]?: U;
};
export declare type FieldValidatorMap<T> = Partial<{
    [P in keyof T]: FieldValidator;
}>;
export declare enum MessageType {
    /**
     * Field or form contains an error. If any validator, will return a message with this
     * code, the form will be treated as invalid.
     */
    ERROR = "error",
    /**
     * Field or form contains a state for what user should be warned. Form can still be
     * be valid event if it contains warnings.
     */
    WARNING = "warning",
    /**
     * Field or form contains a state for what user should be hinted. Form can be treaded valid
     * when containing this kind of messages.
     */
    HINT = "hint",
    /**
     * Error message given, when validator callback fails with the exception. Exception error
     * message will be included into message.
     */
    VALIDATION_ERROR = "validation_error"
}
export declare enum FormValidationMessageCode {
    NONE = 0,
    FIELD_IS_REQUIRED = 1,
    CUSTOM = 100,
    INTERNAL_FIELD_NOT_FOUND = 1000,
    INTERNAL_ERROR = 9999
}
/**
 * Set of message types triggered in validatio
 */
export declare type MessageTypeSet = {
    [P in keyof MessageType]: true;
};
export interface ValidationMessage {
    /**
     * Type of the error message.
     *
     * @type {MessageType}
     * @memberof ValidationMessage
     */
    type: MessageType;
    /**
     * Message code if any available
     */
    code?: FormValidationMessageCode;
    /**
     * The actual message
     *
     * @type {string}
     * @memberof ValidationMessage
     */
    message: string;
}
/**
 * Message field will indicate if the field has same sort of message triggered during the validation. If not, the value should be false
 */
export declare type MessageField = false | ValidationMessage;
/**
 * Set of message fields
 */
export declare type MessageFields<T> = MapType<T, MessageField>;
/**
 * Set of fields with a value filled by the user
 */
export declare type FilledFormFields<T> = Partial<BooleanMap<T>>;
export interface ValidatedForm<T> {
    /**
     * Form values after validation. These values can be altered by the preprocessing.
     */
    values: T;
    /**
     *
     */
    messages: MessageFields<T>;
    /**
     * True if form has no errors after validation.
     */
    isFormValid: boolean;
    /**
     * Message included into form level
     */
    formMessage?: MessageField;
}
export interface RequiredFieldValidationFields<T> {
    /**
     * The actual form values. If there are preprocessing defined in rules, all preprocessing
     * is already done to these values.
     */
    values: T;
    /**
     * Set of fields, where user has already set an value
     */
    filled: FilledFormFields<T>;
    /**
     * Tells if the form is valid after field validators
     */
    isFormValid: boolean;
}
export declare type MessageFunction = () => string;
/**
 * Rule definitions for form validation
 */
export interface FormValidationRules<T> {
    /**
     * Do validation
     */
    validateForm?: (form: RequiredFieldValidationFields<T>) => {
        isFormValid: boolean;
        /**
         * Field specific messages
         */
        messages: MessageFields<T>;
        /**
         * Message for the whole form
         */
        formMessage?: ValidationMessage;
    };
    /**
     * Field specific validators
     */
    fields?: Partial<FieldValidatorMap<T>>;
    /**
     * Default message texts used on validation cases triggered
     * internally by the form validator
     */
    defaultMessages?: {
        /**
         * Message text shown when field is required, but has no value
         */
        requiredField?: string | MessageFunction;
    };
}
/**
 * Options to control how the form validation flow will be executed.
 */
export interface ValidateFormOptions {
    /**
     * Disable preprocessor. Default value is true
     */
    usePreprocessor: boolean;
}
/**
 * Validate form input against a defined rules
 *
 * @export
 * @template T Type of form fields
 *
 * @param form The form info for controlling validation
 * @param options: Extra options to control validation process
 * @returns {ValidatedForm<T>}
 */
export declare function validateForm<T>(form: InputForm<T>, options?: ValidateFormOptions): Form<T>;
/**
 * Checks if form has specific type of messages included.
 *
 * @param response Form response object
 * @param type Message type type be checked
 */
export declare const formHaveMessagesOfType: (response: ValidatedForm<any>, type: MessageType) => boolean;
export {};


lib/index.js
/**
 * # ts-form-validation
 *
 * A simple form validation library for typescript users.
 *
 * @author Ville Venäläinen / Mindhive Oy
 */
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
export var DEFAULT_ERRORMESSAGE_FIELD_IS_REQUIRED = 'This field is required';
/**
 * Initialize form with default set
 */
export var initForm = function (values, rules) { return ({
    values: values,
    filled: {},
    messages: {},
    rules: rules,
    isFormValid: false,
}); };
export var MessageType;
(function (MessageType) {
    /**
     * Field or form contains an error. If any validator, will return a message with this
     * code, the form will be treated as invalid.
     */
    MessageType["ERROR"] = "error";
    /**
     * Field or form contains a state for what user should be warned. Form can still be
     * be valid event if it contains warnings.
     */
    MessageType["WARNING"] = "warning";
    /**
     * Field or form contains a state for what user should be hinted. Form can be treaded valid
     * when containing this kind of messages.
     */
    MessageType["HINT"] = "hint";
    /**
     * Error message given, when validator callback fails with the exception. Exception error
     * message will be included into message.
     */
    MessageType["VALIDATION_ERROR"] = "validation_error";
})(MessageType || (MessageType = {}));
export var FormValidationMessageCode;
(function (FormValidationMessageCode) {
    FormValidationMessageCode[FormValidationMessageCode["NONE"] = 0] = "NONE";
    FormValidationMessageCode[FormValidationMessageCode["FIELD_IS_REQUIRED"] = 1] = "FIELD_IS_REQUIRED";
    FormValidationMessageCode[FormValidationMessageCode["CUSTOM"] = 100] = "CUSTOM";
    FormValidationMessageCode[FormValidationMessageCode["INTERNAL_FIELD_NOT_FOUND"] = 1000] = "INTERNAL_FIELD_NOT_FOUND";
    FormValidationMessageCode[FormValidationMessageCode["INTERNAL_ERROR"] = 9999] = "INTERNAL_ERROR";
})(FormValidationMessageCode || (FormValidationMessageCode = {}));
/**
 * Validate form input against a defined rules
 *
 * @export
 * @template T Type of form fields
 *
 * @param form The form info for controlling validation
 * @param options: Extra options to control validation process
 * @returns {ValidatedForm<T>}
 */
export function validateForm(form, options) {
    if (options === void 0) { options = { usePreprocessor: true }; }
    var values = form.values, filled = form.filled, rules = form.rules;
    var messages = {};
    var isFormValid = true;
    if (rules.fields) {
        for (var fieldKey in rules.fields) {
            var validator = rules.fields[fieldKey];
            if (validator) {
                var field = values[fieldKey];
                // Do preprocessing first
                var value = options.usePreprocessor
                    ? preprocessFormValue(validator, field)
                    : field;
                values[fieldKey] = value;
                // Check if required field
                if (validator.required &&
                    (value === undefined ||
                        (typeof value === 'string' && value.length === 0))) {
                    if (filled[fieldKey]) {
                        messages[fieldKey] = getRequiredFieldErrorMessage(rules, fieldKey);
                    }
                    isFormValid = false;
                    continue;
                }
                // Custom validation
                if (validator.validate) {
                    var validationResult = validator.validate(value);
                    if (validationResult) {
                        if (filled[fieldKey]) {
                            messages[fieldKey] = validationResult;
                        }
                        if (isFormValid) {
                            isFormValid = validationResult.type !== MessageType.ERROR;
                        }
                    }
                }
            }
        }
    }
    if (rules.validateForm) {
        // Erase possible old form message
        var formCopy = __assign({}, form);
        delete formCopy.formMessage;
        var result = rules.validateForm({ isFormValid: isFormValid, values: values, filled: filled });
        if (result.formMessage && result.isFormValid) {
            result.isFormValid = result.formMessage.type !== MessageType.ERROR;
        }
        return __assign({}, formCopy, result, { values: values, messages: __assign({}, messages, result.messages) });
    }
    return __assign({}, form, { values: values,
        messages: messages,
        isFormValid: isFormValid });
}
/**
 * Checks if form has specific type of messages included.
 *
 * @param response Form response object
 * @param type Message type type be checked
 */
export var formHaveMessagesOfType = function (response, type) {
    if (response.messages) {
        for (var key in response.messages) {
            var field = response.messages[key];
            if (field && field.type === MessageType.ERROR) {
                return true;
            }
        }
    }
    return false;
};
function getRequiredFieldErrorMessage(rules, fieldKey) {
    try {
        var fields = rules.fields || {};
        var field = fields[fieldKey];
        if (!field) {
            return {
                type: MessageType.VALIDATION_ERROR,
                code: FormValidationMessageCode.INTERNAL_FIELD_NOT_FOUND,
                message: '',
            };
        }
        var message = field.requiredText ||
            (rules.defaultMessages && rules.defaultMessages.requiredField) ||
            DEFAULT_ERRORMESSAGE_FIELD_IS_REQUIRED;
        if (typeof message === 'function') {
            message = message();
        }
        return {
            type: MessageType.ERROR,
            code: FormValidationMessageCode.FIELD_IS_REQUIRED,
            message: message,
        };
    }
    catch (error) {
        return {
            type: MessageType.VALIDATION_ERROR,
            code: FormValidationMessageCode.INTERNAL_ERROR,
            message: error,
        };
    }
}
/**
 * Do proprocessing for the field value
 *
 * @param {(FieldValidator | undefined)} validator
 * @param {*} value
 * @returns Preprocessed value
 */
function preprocessFormValue(validator, value) {
    if (!value) {
        return value;
    }
    if (validator.trim && typeof value === 'string') {
        value = value.trim();
    }
    if (validator.preprocess) {
        value = validator.preprocess(value);
    }
    return value;
}
//# sourceMappingURL=index.js.map